### PR TITLE
wasmer: Complete and fix v0.8 runtime interface

### PR DESCRIPTION
**Pull request by FlorianFranzen**
## Changes

Updating the gossamer adapters host-api tests I found myself chasing down a trail of bugs, this is the result.

The following changes are included:
 - Adds following (dummy) functions to wasmer backend:
   - ext_storage_exists_version_1 
   - ext_default_child_storage_read_version_1 
   - ext_default_child_storage_exists_version_1
   - ext_default_child_storage_clear_prefix_version_1
   - ext_default_child_storage_next_key_version_1 
   - ext_crypto_ed25519_public_keys_version_1
   - ext_crypto_ed25519_sign_version_1
   - ext_crypto_sr25519_verify_version_1
   - ext_crypto_secp256k1_ecdsa_recover_version_1
   - ext_hashing_twox_256_version_1
   - ext_trie_blake2_256_root_version_1
 - Implements the following bindings for the wasmer backend:
   - ext_hashing_blake2_128_version_1
   - ext_hashing_blake2_256_version_1 
   - ext_hashing_keccak_256_version_1 
   - ext_hashing_twox_64_version_1 
   - ext_hashing_twox_256_version_1 
   - ext_storage_set_version_1
 - Fixes scale encoding of `optional.Bytes` 
 - Implements scale decoding of `optional.Bytes`
 - Fix for the lifetime of memory slices (add runtime.Memory interface)
 - Do not overwrite exported memory in wasmer backend

**It should be noted that the current implementation of `ext_storage_set_version_1` can not distinguish between an empty storage entry and a none existing one, due to a lake of distinction in the gossamer storage api.**

## Tests

```
$ ./runtests.jl gossamer host-api
Test Summary: | Pass  Broken  Total
Host API      |   80     259    339
```
(Crashed and did not pass a single test before.)

## Checklist

- [ ] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [x] I have reviewed my own pull request before requesting a review
- [ ] All integration tests and required coverage checks are passing

## Issues

- #1104
- #1105

Probably also fixes some of the previous memory issues in the wasm environment.

### CODE REVIEW

**Comment by FlorianFranzen** [comment]
It seems the linter dislikes the extra lines of visual separations between HostAPI classes: `File is not `goimports`-ed`

**Review by noot — CHANGES_REQUESTED** [review]
looks good, seems like there are a few lint issues:
```
lib/runtime/wasmer/imports.go:122:6: func `toWasmMemory` is unused (unused)
lib/runtime/wasmer/imports.go:159:6: func `toWasmMemory64` is unused (unused)
lib/runtime/wasmer/imports.go:163:6: func `toWasmMemory128` is unused (unused)
lib/runtime/wasmer/imports.go:146:24: error strings should not be capitalized or end with punctuation or a newline (golint)
		return 0, errors.New("[runtime/wasmer] Internal byte array size missmatch.") 
		                     ^
lib/runtime/wasmer/imports.go:118: File is not `gofmt`-ed with `-s` (gofmt)
	return memory[ptr:ptr+size]
```

could you remove the unused code, update the error string, and run `go fmt ./...`?

**Comment by FlorianFranzen** [comment]
I think the CI fails because I do no have access to your Github secrets from CI jobs I "control":
```
echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
Must provide --username with --password-stdin
The command "echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin" exited with 1.
```

**Comment by FlorianFranzen** [comment]
Ok, I ended up fixing a few more bugs in the end. Should be done for now, as the wasmer backend now passes all test ported from the legacy version.

File: lib/runtime/wasmer/instance.go [hunk 1]
```diff
@@ -29,10 +29,7 @@ import (
 // Name represents the name of the interpreter
 const Name = "wasmer"
 
-var _ runtime.LegacyInstance = (*LegacyInstance)(nil)
```
Comment (noot) [inline lib/runtime/wasmer/instance.go:32]: why is this removed? could you re-add it?

Comment (FlorianFranzen) [reply lib/runtime/wasmer/instance.go:32]: Does that have any effect?

Comment (FlorianFranzen) [reply lib/runtime/wasmer/instance.go:32]: Oh, sorry. Had to learn that these are interface checks. I added it back to the code.

**Review by noot — APPROVED** [review]
looks good, thanks!

unfortunately this still doesn't fix the corrupted state error, you can test it by commenting out the `t.Skip()` in `TestInstance_GrandpaAuthorities_NodeRuntime` in `lib/runtime/wasmer/exports_test.go` for instance. I'm not sure if this was intended to fix that or not?

**Review by arijitAD — APPROVED** [review]
